File: cs/cslox/Parser/Parser.cs
using System;
using cslox.Scanning;

namespace cslox.Parser;

class Parser
{
    private readonly List<Token> tokens;
    private int current = 0;

    public Parser(List<Token> tokens)
    {
        this.tokens = tokens;
    }

    public List<Stmt> Parse()
    {
        var statements = new List<Stmt>();
        while (!IsAtEnd())
        {
            var statement = Declaration();
            if (statement != null)
            {
                statements.Add(statement);
            }
        }

        return statements;
    }

    private Stmt? Declaration()
    {
        try
        {
            if (Match(TokenType.VAR))
            {
                return VarDeclaration();
            }

            return Statement();
        }
        catch (ParseError)
        {
            Synchronize();
            return null;
        }
    }

    private Stmt Statement()
    {
        if (Match(TokenType.FOR))
        {
            return ForStatement();
        }
        if (Match(TokenType.IF))
        {
            return IfStatement();
        }
        if (Match(TokenType.PRINT))
        {
            return PrintStatement();
        }
        if (Match(TokenType.WHILE))
        {
            return WhileStatement();
        }
        if (Match(TokenType.LEFT_BRACE))
        {
            return new Block(Block());
        }

        return ExpressionStatement();
    }

    private Stmt ForStatement()
    {
        Consume(TokenType.LEFT_PAREN, "Expect '(' after 'for'.");

        Stmt? initializer;
        if (Match(TokenType.SEMICOLON))
        {
            initializer = null;
        }
        else if (Match(TokenType.VAR))
        {
            initializer = VarDeclaration();
        }
        else
        {
            initializer = ExpressionStatement();
        }

        Expr? condition = null;
        if (!Check(TokenType.SEMICOLON))
        {
            condition = Expression();
        }
        Consume(TokenType.SEMICOLON, "Expect ';' after loop condition.");

        Expr? increment = null;
        if (!Check(TokenType.RIGHT_PAREN))
        {
            increment = Expression();
        }
        Consume(TokenType.RIGHT_PAREN, "Expect ')' after for clauses.");

        var body = Statement();

        // desugar for loop
        if (increment != null)
        {
            body = new Block(new List<Stmt> { body, new ExpressionStmt(increment) });
        }

        if (condition == null)
        {
            condition = new Literal(true);
        }
        body = new While(condition, body);

        if (initializer != null)
        {
            body = new Block(new List<Stmt> { initializer, body });
        }

        return body;
    }

    private Stmt IfStatement()
    {
        Consume(TokenType.LEFT_PAREN, "Expect '(' after 'if'.");
        var condition = Expression();
        Consume(TokenType.RIGHT_PAREN, "Expect ')' after if condition.");

        var thenBranch = Statement();
        Stmt? elseBranch = null;
        if (Match(TokenType.ELSE))
        {
            elseBranch = Statement();
        }

        return new If(condition, thenBranch, elseBranch);
    }

    private Stmt PrintStatement()
    {
        Expr value = Expression();
        Consume(TokenType.SEMICOLON, "Expect ';' after value.");
        return new Print(value);
    }

    private Stmt WhileStatement()
    {
        Consume(TokenType.LEFT_PAREN, "Expect '(' after 'while'.");
        var condition = Expression();
        Consume(TokenType.RIGHT_PAREN, "Expect ')' after condition.");
        var body = Statement();

        return new While(condition, body);
    }

    private List<Stmt> Block()
    {
        var statements = new List<Stmt>();

        while (!Check(TokenType.RIGHT_BRACE) && !IsAtEnd())
        {
            var statement = Declaration();
            if (statement != null)
            {
                statements.Add(statement);
            }
        }

        Consume(TokenType.RIGHT_BRACE, "Expect '}' after block.");
        return statements;
    }

    private Stmt ExpressionStatement()
    {
        var expr = Expression();
        Consume(TokenType.SEMICOLON, "Expect ';' after expression.");
        return new ExpressionStmt(expr);
    }

    private Stmt VarDeclaration()
    {
        var name = Consume(TokenType.IDENTIFIER, "Expect variable name.");

        Expr? initializer = null;
        if (Match(TokenType.EQUAL))
        {
            initializer = Expression();
        }

        Consume(TokenType.SEMICOLON, "Expect ';' after variable declaration.");
        return new Var(name, initializer);
    }

    private Expr Expression()
    {
        return Assignment();
    }

    private Token Peek() => tokens[current];
    private Token Previous() => tokens[current - 1];

    private bool IsAtEnd() => Peek().Type == TokenType.EOF;

    private Token Advance()
    {
        if (!IsAtEnd())
        {
            current++;
        }

        return Previous();
    }

    private bool Check(TokenType tokenType)
    {
        if (IsAtEnd())
        {
            return false;
        }
        return Peek().Type == tokenType;
    }

    private bool Match(params TokenType[] tokenTypes)
    {
        foreach (var tokenType in tokenTypes)
        {
            if (Check(tokenType))
            {
                Advance();
                return true;
            }
        }

        return false;
    }

    // assignment     → IDENTIFIER "=" assignment
    //                | logic_or ;
    private Expr Assignment()
    {
        var expr = Or();

        if (Match(TokenType.EQUAL))
        {
            var equals = Previous();
            var value = Assignment();

            if (expr is Variable)
            {
                var name = ((Variable)expr).Name;
                return new Assign(name, value);
            }
        }

        return expr;
    }

    // logic_or       → logic_and ( "or" logic_and )* ;
    private Expr Or()
    {
        var expr = And();
        while (Match(TokenType.OR))
        {
            var @operator = Previous();
            var right = And();
            expr = new Logical(expr, @operator, right);
        }

        return expr;
    }

    // logic_and      → equality ( "and" equality )* ;
    private Expr And()
    {
        var expr = Equality();

        while (Match(TokenType.AND))
        {
            var @operator = Previous();
            var right = Equality();
            expr = new Logical(expr, @operator, right);
        }

        return expr;
    }

    // equality       → comparison ( ( "!=" | "==" ) comparison )* ;
    private Expr Equality()
    {
        var expr = Comparison();

        while (Match(TokenType.BANG_EQUAL, TokenType.EQUAL_EQUAL))
        {
            var @operator = Previous();
            var right = Comparison();
            expr = new Binary(expr, @operator, right);
        }

        return expr;
    }

    // comparison     → term(( ">" | ">=" | "<" | "<=" ) term )* ;
    private Expr Comparison()
    {
        var expr = Term();

        while (Match(TokenType.GREATER, TokenType.GREATER_EQUAL, TokenType.LESS, TokenType.LESS_EQUAL))
        {
            var @operator = Previous();
            var right = Term();
            expr = new Binary(expr, @operator, right);
        }

        return expr;
    }

    // term           → factor ( ( "-" | "+" ) factor )* ;
    private Expr Term()
    {
        var expr = Factor();

        while (Match(TokenType.MINUS, TokenType.PLUS))
        {
            var @operator = Previous();
            var right = Factor();
            expr = new Binary(expr, @operator, right);
        }

        return expr;
    }

    // factor         → unary ( ( "/" | "*" ) unary )* ;
    private Expr Factor()
    {
        var expr = Unary();

        while (Match(TokenType.SLASH, TokenType.STAR))
        {
            var @operator = Previous();
            var right = Unary();
            expr = new Binary(expr, @operator, right);
        }

        return expr;
    }

    // unary          → ( "!" | "-" ) unary
    //                | primary ;
    private Expr Unary()
    {
        if (Match(TokenType.BANG, TokenType.MINUS))
        {
            var @operator = Previous();
            var right = Unary();
            return new Unary(@operator, right);
        }

        return Call();
    }

    // call           → primary ( "(" arguments? ")" )* ;
    private Expr Call()
    {
        var expr = Primary();

        while (true)
        {
            if (Match(TokenType.LEFT_PAREN))
            {
                expr = FinishCall(expr);
            }
            else
            {
                break;
            }
        }

        return expr;
    }

    private Expr FinishCall(Expr callee)
    {
        var arguments = new List<Expr>();
        if (!Check(TokenType.RIGHT_PAREN))
        {
            do
            {
                if (arguments.Count >= 255)
                {
                    Error(Peek(), "Can't have more than 255 arguments.");
                }
                arguments.Add(Expression());
            } while (Match(TokenType.COMMA));
        }

        var paren = Consume(TokenType.RIGHT_PAREN, "Expect ')' after arguments.");

        return new Call(callee, paren, arguments);
    }

    // primary        → NUMBER | STRING | "true" | "false" | "nil"
    //                | "(" expression ")" ;
    private Expr Primary()
    {
        if (Match(TokenType.FALSE))
        {
            return new Literal(false);
        }
        if (Match(TokenType.TRUE))
        {
            return new Literal(true);
        }
        if (Match(TokenType.NIL))
        {
            return new Literal(null);
        }

        if (Match(TokenType.NUMBER, TokenType.STRING))
        {
            return new Literal(Previous().Literal);
        }

        if (Match(TokenType.IDENTIFIER))
        {
            return new Variable(Previous());
        }

        if (Match(TokenType.LEFT_PAREN))
        {
            var expr = Expression();
            Consume(TokenType.RIGHT_PAREN, "Expect ')' after expression.");
            return new Grouping(expr);
        }

        throw Error(Peek(), "Expect expression.");
    }

    private Token Consume(TokenType tokenType, string message)
    {
        if (Check(tokenType))
        {
            return Advance();
        }

        throw Error(Peek(), message);
    }

    private ParseError Error(Token token, string message)
    {
        Lox.Error(token, message);
        return new ParseError();
    }

    private void Synchronize()
    {
        Advance();

        while (!IsAtEnd())
        {
            if (Previous().Type == TokenType.SEMICOLON)
            {
                return;
            }

            switch (Peek().Type)
            {
                case TokenType.CLASS:
                case TokenType.FUN:
                case TokenType.VAR:
                case TokenType.FOR:
                case TokenType.IF:
                case TokenType.WHILE:
                case TokenType.PRINT:
                case TokenType.RETURN:
                    return;
            }
        }

        Advance();
    }
}

class ParseError : Exception
{
}

File: tree-walk-interpreter/go/internal/parser/parser.go
package parser

import (
	loxError "github.com/kashifsoofi/go-lox/internal/error"
	"github.com/kashifsoofi/go-lox/internal/scanner"
)

var (
	tokens  []*scanner.Token
	current int
)

func Parse(scannedTokens []*scanner.Token) Expr {
	tokens = scannedTokens
	current = 0

	return parse()
}

func parse() Expr {
	var expr Expr
	defer func() {
		if err := recover(); err != nil {
			if _, ok := err.(ParseError); ok {
				expr = nil
				return
			}
			panic(err)
		}
	}()

	expr = expression()
	return expr
}

func expression() Expr {
	return equality()
}

func equality() Expr {
	expr := comparison()

	for match(scanner.BangEqual, scanner.EqualEqual) {
		operator := previous()
		right := comparison()
		expr = NewBinary(expr, operator, right)
	}

	return expr
}

func comparison() Expr {
	expr := term()

	for match(scanner.Greater, scanner.GreaterEqual, scanner.Less, scanner.LessEqual) {
		operator := previous()
		right := term()
		expr = NewBinary(expr, operator, right)
	}

	return expr
}

func term() Expr {
	expr := factor()

	for match(scanner.Minus, scanner.Plus) {
		operator := previous()
		right := factor()
		expr = NewBinary(expr, operator, right)
	}

	return expr
}

func factor() Expr {
	expr := unary()

	for match(scanner.Slash, scanner.Star) {
		operator := previous()
		right := unary()
		expr = NewBinary(expr, operator, right)
	}

	return expr
}

func unary() Expr {
	if match(scanner.Bang, scanner.Minus) {
		operator := previous()
		right := unary()
		return NewUnary(operator, right)
	}

	return primary()
}

func primary() Expr {
	if match(scanner.False) {
		return NewLiteral(false)
	}
	if match(scanner.True) {
		return NewLiteral(true)
	}
	if match(scanner.Nil) {
		return NewLiteral(nil)
	}

	if match(scanner.Number, scanner.String) {
		return NewLiteral(previous().Literal)
	}

	if match(scanner.LeftParen) {
		expr := expression()
		consume(scanner.RightParen, "Expect ')' after expression.")
		return NewGrouping(expr)
	}

	panic(error(peek(), "Expect expression."))
}

func match(tokenTypes ...scanner.TokenType) bool {
	for _, tokenType := range tokenTypes {
		if check(tokenType) {
			advance()
			return true
		}
	}

	return false
}

func check(tokenType scanner.TokenType) bool {
	if isAtEnd() {
		return false
	}

	return peek().Type == tokenType
}

func advance() *scanner.Token {
	if !isAtEnd() {
		current++
	}
	return previous()
}

func isAtEnd() bool {
	return peek().Type == scanner.EOF
}

func peek() *scanner.Token {
	return tokens[current]
}

func previous() *scanner.Token {
	return tokens[current-1]
}

func consume(tokenType scanner.TokenType, message string) *scanner.Token {
	if check(tokenType) {
		return advance()
	}

	panic(error(previous(), message))
}

func error(token *scanner.Token, message string) ParseError {
	if token.Type == scanner.EOF {
		loxError.ErrorWithWhere(token.Line, "at end", message)
	} else {
		loxError.ErrorWithWhere(token.Line, "at '"+token.Lexeme+"'", message)
	}
	return NewParseError(token, message)
}

func synchronize() {
	advance()

	for !isAtEnd() {
		if previous().Type == scanner.Semicolon {
			return
		}

		switch peek().Type {
		case scanner.Class:
			return
		case scanner.Fun:
			return
		case scanner.Var:
			return
		case scanner.For:
			return
		case scanner.If:
			return
		case scanner.While:
			return
		case scanner.Print:
			return
		case scanner.Return:
			return
		}

		advance()
	}
}
